New module for JPL Spectroscopy Catalog

gianninapr opened the pull request:
[JPL Spectroscopy Catalog](https://spec.jpl.nasa.gov/ftp/pub/catalog/catform.html) 

This is a new module to query data from the catalog above. It follows the browser form template and includes the full functionality of the form. This module can also match the name of the molecules using user-defined regular expressions. What makes this query unique is that the catalog includes each molecule's partition function for different temperatures, an example of how to access these values is shown in the documentation of the module. 

[Link to the issue in StackOverflow](https://stackoverflow.com/questions/50822755/https-post-to-access-catalog-with-python-requests). This issue has been resolved with the latest commit, using an ordered list of tuples for the payload instead of a dictionary due to compatibility issues with the catalog's server and python requests. 

A changelog has been added. 

Aside from being functional for personal usage, this module will also work as a service to the bigger NASA-funded project (sbpy). More information about the services that will be added to astroquery as part of this project can be seen in issue #1075

pep8speaks commented:
Hello @gianninapr! Thanks for submitting the PR.



 - In the file [`astroquery/jplspec/core.py`](https://github.com/astropy/astroquery/blob/af97085510ab950b234eaa6e4feb7bfbe7fc9a56/astroquery/jplspec/core.py), following are the PEP8 issues :

> [Line 108:69](https://github.com/astropy/astroquery/blob/af97085510ab950b234eaa6e4feb7bfbe7fc9a56/astroquery/jplspec/core.py#L108): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','



 - In the file [`astroquery/jplspec/setup_package.py`](https://github.com/astropy/astroquery/blob/af97085510ab950b234eaa6e4feb7bfbe7fc9a56/astroquery/jplspec/setup_package.py), following are the PEP8 issues :

> [Line 6:1](https://github.com/astropy/astroquery/blob/af97085510ab950b234eaa6e4feb7bfbe7fc9a56/astroquery/jplspec/setup_package.py#L6): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

migueldvb commented:
We will need to provide ordered data in the post request to the JPL catalog for the query to work. Is there any preference in using a list of tuples for the data argument (not documented in the docstring but it is alluded to [here](http://docs.python-requests.org/en/master/user/quickstart/#more-complicated-post-requests)) or an `OrderedDict`?

keflavich commented:
There are many commits in here that do not belong to this branch, afaict.  Consider rebasing?

gianninapr commented:
@keflavich we've fixed the conflict and rebased. The payload of the query as of now is an ordered list due to the previous conflict with the server, to have the payload as an ordered list instead of a dictionary is the only solution we found as of yet. We'll remain in contact with the JPL team and if they fix the issue we will restore the payload as a dictionary.

keflavich commented on astroquery/jplspec/core.py:
```diff
@@ -0,0 +1,253 @@
+
+# all Query classes should inherit from BaseQuery.
+from ..query import BaseQuery
+# async_to_sync generates the relevant query tools from _async methods
```
these comments are from the template and are intended to be instructional; could you remove them?  I leave to your discretion which ones ought to be removed

keflavich commented on astroquery/jplspec/core.py:
```diff
@@ -0,0 +1,253 @@
+        if len(result) > self.maxlines:
+            warnings.warn("This form is currently limited to {0} lines.\n\
+                          Please limit your search.".format(
+
```
blank line here should be removed?

keflavich commented on astroquery/jplspec/tests/test_jplspec.py:
```diff
@@ -0,0 +1,120 @@
+            return f.read()
+
+
+def test_input_async():
```
As written, this looks like it's a remote query but is not tagged with `@remote_data`.  It should either be mocked so that the query is performed locally or marked `remote`.  A remote query that is not marked remote should cause the tests to fail.

keflavich replied:
I think the intent here is to have these mocked, though.  You should follow examples from other modules, e.g. https://github.com/astropy/astroquery/blob/master/astroquery/vizier/tests/test_vizier.py#L29, to see the `pytest.fixture` mocking in action.

gianninapr replied:
We had not tagged it as remote since the get_query_payload=True bypasses the POST request and only returns the payload of the query which is the dictionary of values entered by the user, it doesn't actually make the query. We made these tests to test that the function was accurately passing the values to the dictionary that goes into the request. We mocked all the other tests because these actually require the query to  be made. We tested the script locally without --remote-data and they passed. Would we still have to mock this?

keflavich commented on docs/jplspec/jplspec.rst:
```diff
@@ -0,0 +1,256 @@
+.. code-block:: python
+
+   >>> response = JPLSpec.query_lines(min_frequency=100 * u.GHz,
+                                            max_frequency=1000 * u.GHz,
```
make the whitespace here match the previous line

keflavich commented on docs/jplspec/jplspec.rst:
```diff
@@ -0,0 +1,256 @@
+   {'MinNu': 100.0, 'MaxNu': 1000.0, 'Mol': '28001 CO', 'UnitNu': 'GHz',
+   'StrLim': -500, 'MaxLines': 7}
+
+The units of the columns can be displayed by calling ``response.info``. These
```
show a call to `response.info` here

keflavich commented on docs/jplspec/jplspec.rst:
```diff
@@ -0,0 +1,256 @@
+
+***********************************************
+JPL Spectroscopy Queries (astroquery.jplspec)
+***********************************************
```
the number of *'s should match the number of characters in the line

keflavich commented on docs/jplspec/jplspec.rst:
```diff
@@ -0,0 +1,256 @@
+        'HCOCH2OH': 60006,
+        'NH2CH2CH2OH': 61004}
+
+As you can see, the 'H20' string was processed as a regular expression,
```
H2O not H20

keflavich commented on docs/jplspec/jplspec.rst:
```diff
@@ -0,0 +1,256 @@
+   >>> {'H2O': 18003}
+
+
+As seen above, the regular expression "H2O$" yields only an exact match.
```
add "because the special character `$` matches the end of the line."

gianninapr commented:
@keflavich Thank you for the line by line inputs. I have taken care of them and already committed the changes. Let me know if anything else is needed from me or if I misread any of your comments and didn't fully satisfy the requests you made.

bsipocz commented on astroquery/jplspec/core.py:
```diff
@@ -0,0 +1,246 @@
+        min_strength : int, optional
+            Minimum strength in catalog units
+        max_lines :  int, optional
+            Maximum number of lines to query, the default is 1000.
```
this default is in contradiction with the signature. Could you please revise the defaults (and preferably list them for the other parameters, too here in the docstring)

bsipocz commented on astroquery/jplspec/core.py:
```diff
@@ -0,0 +1,246 @@
+        --------
+        >>> table = JPLSpec.query_lines(min_frequency=100*u.GHz,
+        ...                             max_frequency=200*u.GHz,
+        ...                             min_strength=-500, molecule=18003) # doctest: +SKIP
```
could you maybe try and run this as a remote doctest? Rather than `+SKIP` it's called `+REMOTE_DATA`

bsipocz commented on astroquery/jplspec/core.py:
```diff
@@ -0,0 +1,246 @@
+                            format='fixed_width')
+
+        if len(result) > self.maxlines:
+            warnings.warn("This form is currently limited to {0} lines.\n\
```
no need for the last `\`, you can close the line with a quotation mark and start again on the next one, it will automatically join the two strings

bsipocz commented on astroquery/jplspec/core.py:
```diff
@@ -0,0 +1,246 @@
+
+        Parameters
+        -----------
+        catfile : default 'catdir.cat'
```
str, name of the file, rather that a file object (I mean it's obvious as the default is a string, but can be made more consistent).

bsipocz commented on docs/jplspec/jplspec.rst:
```diff
@@ -0,0 +1,311 @@
+   >>> plt.show()
+
+
+.. figure:: images/docplot_jplspec.png
```
actually it would be nice to let the docs generate the figure. However I suggest to do that in a follow-up PR rather than holding this up from merging.

gianninapr replied:
No problem, I can definitely work on this for a follow up PR as you mentioned.

gianninapr commented:
@bsipocz thank you for the pointers! I have committed the changes you requested. I can definitely work on running doctests for a follow up PR. Let me know if anything else is needed of me!

bsipocz commented:
Thanks @gianninapr!

 I'm fairly certain that this merge will trigger a deployment to pypi, so you'll the next version (installed with `pip install --upgrade --pre astroquery`) will have this new feature.
